crates/services/p2p/src/codecs/bounded.rs

```diff
@@ -0,0 +1,156 @@
+use std::{
+    io,
+    marker::PhantomData,
+};
+
+use async_trait::async_trait;
+use futures::{
+    AsyncRead,
+    AsyncReadExt,
+    AsyncWriteExt,
+};
+use libp2p::request_response;
+use strum::IntoEnumIterator as _;
+
+use crate::request_response::{
+    messages::{
+        RequestMessage,
+        V1ResponseMessage,
+        V2ResponseMessage,
+    },
+    protocols::RequestResponseProtocol,
+};
+
+use super::{
+    DataFormatCodec,
+    RequestResponseProtocols,
+};
+
+#[derive(Debug, Clone)]
+pub struct BoundedCodec<Format> {
+    _format: PhantomData<Format>,
+    /// Used for `max_size` parameter when reading Response Message
+    /// Necessary in order to avoid DoS attacks
+    /// Currently the size mostly depends on the max size of the Block
+    max_response_size: usize,
+}
+
+impl<Format> BoundedCodec<Format> {
+    pub fn new(max_block_size: usize) -> Self {
+        assert_ne!(
+            max_block_size, 0,
+            "PostcardCodec does not support zero block size"
+        );
+
+        Self {
+            _format: PhantomData,
+            max_response_size: max_block_size,
+        }
+    }
+}
+
+/// Since Postcard does not support async reads or writes out of the box
+/// We prefix Request & Response Messages with the length of the data in bytes
+/// We expect the substream to be properly closed when response channel is dropped.
+/// Since the request protocol used here expects a response, the sender considers this
+/// early close as a protocol violation which results in the connection being closed.
+/// If the substream was not properly closed when dropped, the sender would instead
+/// run into a timeout waiting for the response.
+#[async_trait]
+impl<Format> request_response::Codec for BoundedCodec<Format>
```

Comment:
The naming feels off here. A trait named `Codec` should have encode/decode or serialize/deserialize methods. This trait contains read/write methods and seem to be representing primarily I/O operations. I get that this comes from p2p, but perhaps we should find another name for `BoundedCodec`. Right now I'd expect some sort of similarity between the `BoundedCodec` and `UnboundedCodec` types, but they feel very different in what they do with the current implementations.

Perhaps I'm just not understanding this well enough right now. I'd love to hear your thoughts on this.

Reply:
Agreed, I had a lot of troubles trying to find the right terminology (I also did not like `BoundedCodec` and `UnboundedCodec`. In the last version I have made the following changes: 
* `DataFormatCodec` has been replaced by `Encode` and `Decode` traits, 
* `BoundedCodec` is now a `RequestResponseMessageHandler`. I still don't like the Handler part in the name as it might make people think that it's something akin to a task, but I could not find anything better. The `data_format` field has been renamed to `codec`
* `UnboundedCodec` is now `GossipsubMessageHandler`,  and the `data_format` field has been renamed to `codec`, 
* The `Format` type parameter has been renamed to `codec`.
* I did not rename the `GossipsubCodec` trait, suggestions for better names are very welcome.

Reply:
Commits are ead761a08a1a103bdd103f8f7e66da077fb92186 and c5aa3622bb56ef69ce7aee44231ae20a9d05d2c3